md5sum.cpp

```diff
@@ -0,0 +1,140 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software Foundation,
+// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+//
+// http://savannah.nongnu.org/projects/lmi
+// email: <gchicares@sbcglobal.net>
+// snail: Chicares, 186 Belle Woods Drive, Glastonbury CT 06033, USA
+
+#include "pchfile.hpp"
+
+#include "assert_lmi.hpp"
+#include "md5.hpp"
+#include "md5sum.hpp"
+#include "miscellany.hpp"
+
+#include <fstream>
+#include <memory>
+
+std::vector<md5sum_for_file> md5_read_checksum_file(fs::path const& filename)
+{
+    auto const filename_string = filename.string();
+
+    std::ifstream is(filename_string);
+    if(!is)
+        {
+        std::ostringstream oss;
+        oss << "'" << filename_string << "': no such file or directory";
+        throw std::runtime_error(oss.str());
+        }
+
+    std::vector<md5sum_for_file> result_vec;
+
+    int line_number = 0;
+    auto error_with_message = [&line_number, filename_string] (char const* message)
+        {
+        std::ostringstream oss;
+        oss << "'" << filename_string << "': " << message << " at line " << line_number;
+        return oss.str();
+        };
+
+    std::string line;
+    while(std::getline(is, line))
+        {
+        ++line_number;
+
+        // The minimal length: 32(md5sum) + 2(spaces) + 1(shortest file name)
+        size_t const minimal_line_length = 35;
+
+        // Ignore comment lines, which begin with a '#' character.
+        if(0 == line.find_first_of('#'))
+            {
+            continue;
+            }
+
+        if(line.size() < minimal_line_length)
+            {
+            throw std::runtime_error(error_with_message("line too short"));
+            }
+
+        auto const space_pos = line.find_first_of(' ');
+        if(line.size() - 1 <= space_pos)
+            {
+            throw std::runtime_error(error_with_message("incorrect checksum line format"));
+            }
+
+        char const second_delimiter = line[space_pos + 1];
+        if(second_delimiter != ' ' && second_delimiter != '*')
+            {
+            throw std::runtime_error(error_with_message("incorrect checksum line format"));
+            }
+
+        std::string md5sum = line.substr(0, space_pos);
+        std::string file = line.substr(space_pos + 2);
+
+        if(md5sum.size() != chars_per_formatted_hex_byte * md5len)
+            {
+            throw std::runtime_error(error_with_message("incorrect MD5 sum format"));
+            }
+
+        result_vec.emplace_back(md5sum_for_file{file, md5sum});
+        }
+
+    return result_vec;
+}
+
+std::string md5_calculate_checksum(fs::path const& filename)
+{
+    auto const filename_string = filename.string();
+
+    std::vector<unsigned char> md5(md5len);
+
+    std::unique_ptr<std::FILE, decltype(&fclose)> const file
```

Comment:
```suggestion
    std::unique_ptr<std::FILE, decltype(&std::fclose)> const file
```
It's just inconsistent to use for `FILE` and `fopen()` but not `fclose()`.